### ISSUE TITLE
Latency Measurement

### ISSUE THREAD

**Issue by kkeono2** [wujx2001/QwT#1]
Hi, Thank you for your great works.

I was wondering if you could please specify how to implement results in Table 4.

As below Table, i measured latency on original DeiT-Small and quantized one with TensorRT on Orin Agx 64G and profile the results, where i didn't use percentile[23] method. 

<img width="1599" height="288" alt="Image" src="https://github.com/user-attachments/assets/e9c85537-8e74-401f-b651-d157c200f696" />

Thank you.

**Comment by heekhero** [comment]
@kkeono2 Hi there, thanks for trying out QwT!

Table 2 was produced as follows):

**1. Define & Quantize the Model**
We build both the PTQ and QwT‑enhanced models by replacing standard layers with [pytorch_quantization](https://docs.pytorch.org/docs/stable/quantization.html)’s QuantConv2d and QuantLinear.

**2. Export to ONNX**
Once your quantized models are ready (both PTQ only and PTQ + QwT), simply call PyTorch’s ONNX exporter:

We take deit_small as an example.
```python
import torch
# q_model_ptq and q_model_qwt are both on cuda, in eval mode
dummy = torch.randn(64, 3, 224, 224, device='cuda')
torch.onnx.export(q_model_ptq, dummy, 'deit_small_bs_64_ptq.onnx', opset_version=13)
torch.onnx.export(q_model_qwt, dummy, 'deit_small_bs_64_qwt.onnx', opset_version=13)
```
You’ll end up with three ONNX files for batch size 64:

- deit_small_bs_64_fp32.onnx (the full‑precision baseline)

- deit_small_bs_64_ptq.onnx

- deit_small_bs_64_qwt.onnx

**3. Build TRT Engines & Profile**
Use trtexec to build and time each model (all at batch size 64):

```bash
# FP32 baseline
trtexec \
  --onnx=deit_small_bs_64_fp32.onnx \
  --saveEngine=deit_small_fp32.trt \
  --fp32 \

# INT8 PTQ
trtexec \
  --onnx=deit_small_bs_64_ptq.onnx \
  --saveEngine=deit_small_ptq.trt \
  --best \

# INT8 PTQ + QwT
trtexec \
  --onnx=deit_small_bs_64_qwt.onnx \
  --saveEngine=deit_small_qwt.trt \
  --best \
```
- `--fp32`  disables FP16/INT8 and runs everything in FP32 only.
- `--best` enables all supported precisions (FP32, FP16, INT8) and automatically picks the fastest kernels to achieve optimal performance.


**4. Extract Latency Metric**
In each trtexec log, read the **GPU Compute Time** (the pure device‑side kernel execution cost) as the network inference latency.

That gives you the numbers reported in Table 2. 

Now we’re cleaning up and packaging the full latency‑testing scripts—expect the code to be open‑sourced in the next few days. 

Let me know if anything is unclear or you run into issues!

**Comment by kkeono2** [comment]
Thank you for rapid reply and the detailed explanation.

However, I didn't know how to convert QuantLinear and QuantConv2d to [pytorch_quantization](https://docs.pytorch.org/docs/stable/quantization.html)’s QuantConv2d and QuantLinear, where i confused whether to use PTQ-dynamic or PTQ-static.

Would you mind offering the implementation to replace q_model with q_model_ptq and q_model_qwt ?

Thanks.

**Comment by heekhero** [comment]
Hi @kkeono2,

We’ve now open‑sourced the full code, including both the quantized network definitions and the latency‑testing scripts. Please take a look at the **QwT‑cls‑latency‑test** folder:

- **[quant_swin_transformer.py](https://github.com/wujx2001/QwT/blob/main/QwT-cls-latency-test/models/quant_swin_transformer.py)**  
  Defines the quantized Swin Transformer, replacing standard `Conv2d`/`Linear` with `pytorch_quantization`’s `QuantConv2d` and `QuantLinear`.

- **[quant_vision_transformer.py](https://github.com/wujx2001/QwT/blob/main/QwT-cls-latency-test/models/quant_vision_transformer.py)**  
  Shows the Vision Transformer variant with the same quantization wrappers.

In these files you can see how `q_model_ptq` (baseline PTQ) and `q_model_qwt` (PTQ + QwT) are constructed. Calibration is performed directly in the training script: [qwt_vit_and_deit_test_latency.py](https://github.com/wujx2001/QwT/blob/main/QwT-cls-latency-test/qwt_vit_and_deit_test_latency.py).

Feel free to dive in and let me know if you have any further questions!

**Comment by kkeono2** [comment]
I really appreciate you offering the implementation 👍

**Comment by wujx2001** [comment]
Since the issue has been resolved, it is now closed.
